Remove the first migration (WIP)

mjankowski opened the pull request:
As the [prophecy foretold](https://github.com/mastodon/mastodon/pull/27514#issuecomment-1783165125) this is a pared down version of https://github.com/mastodon/mastodon/pull/27514#issuecomment-1783165125 which removes only the first-ever migration and attempts to preserve the CI tests around migrations / data integrity. Removing just the first one is sort of silly, so if we like this direction I'll add some. This is mostly for review of alternate approaches here.

mjankowski committed:
Remove the first migration

ClearlyClaire commented on .github/workflows/test-migrations-two-step.yml:
```diff
@@ -77,6 +77,13 @@ jobs:
       - name: Create database
         run: './bin/rails db:create'
 
+      - name: Restore DB as of `20160220174730`
+        run: |
+          mv db/schema.rb db/schema-latest.rb
+          cp db/schema/snapshots/20160220174730.rb db/schema.rb
+          ./bin/rails db:schema:load
+          mv -f db/schema-latest.rb db/schema.rb
```
The following should work:
```suggestion
      - name: Restore DB as of `20160220174730`
        run: |
          SCHEMA=db/schema/snapshots/20160220174730.rb ./bin/rails db:schema:load
```

mjankowski replied:
Ah! Thank you -- I swore this feature existed but could not find it documented, but it's definitely there -- https://github.com/rails/rails/blob/7-1-stable/activerecord/lib/active_record/tasks/database_tasks.rb#L428

That approach is much cleaner, will update PR.

mjankowski replied:
Updated PR to use `SCHEMA` env var to load. Much cleaner.

mjankowski committed:
Simplify schema load

mjankowski commented:
Another approach here would be - instead of deleting all the old migrations, delete all of them up to the very last one we want to delete, and then update that one instead to contain the full contents of the schema at that point in time. Approach described in item 3 here - https://www.globalapptesting.com/engineering/database-migration-tips-in-rails-part-1

This would preserve the ability to just run a normal db:migrate, and we could add comments in that migration to describe what had happened.

ClearlyClaire commented:
> Another approach here would be - instead of deleting all the old migrations, delete all of them up to the very last one we want to delete, and then update that one instead to contain the full contents of the schema at that point in time. Approach described in item 3 here - https://www.globalapptesting.com/engineering/database-migration-tips-in-rails-part-1
> 
> This would preserve the ability to just run a normal db:migrate, and we could add comments in that migration to describe what had happened.

That only sounds workable when starting from a blank database, or am I missing something?

mjankowski commented:
> That only sounds workable when starting from a blank database, or am I missing something?

Yes, that's correct - it's another alternate approach and it would be to preserve the ability to `rails db:migrate` from a freshly created database without relying on the "snapshot" approach currently in this PR.

Hypothetically -- if we wanted to delete migrations 1 through 100, we'd instead replace migration number 100 with the contents of the schema at that point in time. So you could both a) `db:migrate` from scratch, b) `db:schema:load` from scratch (or `db:reset`, etc) -- and we'd be able to preserve the ability to run the CI tasks here (at least for whichever tasks were running on migrations/periods chronologically after the latest deleted migration).

I guess I'm curious:

- How are you feeling about this PR approach vs the previous in #27514 ?
- Do we like EITHER of these sufficiently enough to try to refine them for merge?
- Do we want to see a third approach like what I just described?
- Does this all just feel too brittle and we should just commit to preserving/maintaining/updating the full migration history indefinitely? (or at least for near term future until it becomes untenable)

ClearlyClaire commented:
I like the approach we're having in this PR, but… I've just thought of a couple related issues with removing migration files: forks and downgrades.

- forks: some Mastodon forks have some additional migrations and still allow coming from Mastodon, and they'd need to keep their old custom migrations for that purpose, but it makes things messier
- downgrades: we don't support downgrades, and usually people trying to downgrade end up with a schema that isn't suited for the old code they are running, so this wouldn't be a completely new issue, but this may cause further database weirdness by running very old migrations that we discarded between updates

mjankowski commented:
> I like the approach we're having in this PR

I think I like this PR better than the original which removed the CI tasks ... but I'm sort of on the fence as to whether this is all net improvement over just maintaining the migrations more proactively.

> * forks: some Mastodon forks have some additional migrations and still allow coming from Mastodon, and they'd need to keep their old custom migrations for that purpose, but it makes things messier

Yeah that could get messy ... esp if they had custom migrations critical to running the fork which were in the time period we deleted. They'd have to rebase, and then also modify either the snapshot or new-earliest-migration depending which approach we took. Probably do-able, but certainly more effort.

> * downgrades: we don't support downgrades, and usually people trying to downgrade end up with a schema that isn't suited for the old code they are running, so this wouldn't be a completely new issue, but this may cause further database weirdness by running very old migrations that we discarded between updates

No idea how common that is but certainly makes it more challenging than current. I think the idea of "support rollback of very recent migrations to handle bad deploys" is a good thing to keep intact (which we'd be doing by keeping some reasonable time range on when we deleted). Not sure how reasonable it is to support more arbitrary stuff beyond that. We already have some migrations which are not reversible, so it seems like going backward in time past any of those would already require someone to put their DBA hat on to go earlier.

ClearlyClaire commented:
> I think I like this PR better than the original which removed the CI tasks ... but I'm sort of on the fence as to whether this is all net improvement over just maintaining the migrations more proactively.

I'm in the same boat.

> Not sure how reasonable it is to support more arbitrary stuff beyond that. We already have some migrations which are not reversible, so it seems like going backward in time past any of those would already require someone to put their DBA hat on to go earlier.

Yeah, it's not really about supporting rollbacks: we currently generally don't, and people who do what I described generally don't attempt to properly rollback anyway. My point was more to avoid making things more broken by potentially re-running old migrations.

mjankowski commented:
> > I think I like this PR better than the original which removed the CI tasks ... but I'm sort of on the fence as to whether this is all net improvement over just maintaining the migrations more proactively.
> 
> I'm in the same boat.

Having sat on this for a few days, I think I'm coming down on just not doing anything with this right now.

What I'd like to do instead is pursue some of the other maintenance/improvement ideas (along the lines of https://github.com/mastodon/mastodon/pull/26890 and similar - not just lint/cop fixes, but general improvement of the full migration history). If we get down that path and it's not enough we can revisit something like this PR. If we get down that path and it's all working well, we could formalize the "just preserve/maintain the whole history" as an actual policy and not just an accident.
